{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "f6c5538e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 특정 시트의 특정 컬럼 읽기\n",
    "df = pd.ExcelFile('sample_data/Master_table_SS2241089.xlsx')\n",
    "\n",
    "panel = ''\n",
    "\n",
    "\n",
    "clinical_information = df.parse('clinical_information', header=None)\n",
    "NGS_QC = df.parse('NGS_QC', header=None)\n",
    "SNV = df.parse('SNV')\n",
    "CNV = df.parse('CNV')\n",
    "CNVarm = df.parse('CNVarm')\n",
    "CNV_allFC = df.parse('CNV_allFC')\n",
    "LR_BRCA = df.parse('LR_BRCA')\n",
    "Fusion = df.parse('Fusion', header=1)\n",
    "Splice = df.parse('Splice')\n",
    "IO = df.parse('IO', header=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e207ebd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 검체 정보 파트\n",
    "clinical_dict = dict(zip(clinical_information.iloc[:, 0], clinical_information.iloc[:, 1]))\n",
    "검체_정보 = {\n",
    "    \"검체 정보\": clinical_dict[\"병리번호\"],\n",
    "    \"성별\": clinical_dict[\"성별\"],\n",
    "    \"나이\": clinical_dict[\"나이\"],\n",
    "    \"Unit NO.\": clinical_dict[\"Unit NO.\"],\n",
    "    \"환자명\": clinical_dict[\"환자명\"],\n",
    "    \"채취 장기\": clinical_dict[\"채취 장기\"],\n",
    "    \"원발 장기\": clinical_dict[\"원발 장기\"],\n",
    "    \"의뢰의\": clinical_dict[\"의뢰의\"],\n",
    "    \"의뢰의 소속\": clinical_dict[\"의뢰의 소속\"],\n",
    "    \"검체 유형\": clinical_dict[\"검체 유형\"],\n",
    "    \"검체의 적절성여부\": clinical_dict[\"검체의 적절성\"],\n",
    "    \"검체접수일\": clinical_dict[\"검체접수일\"],\n",
    "    \"결과보고일\": clinical_dict[\"Report date\"],\n",
    "\n",
    "}\n",
    "검체_정보"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4fe9c73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('MET exon14 skipping',\n",
       " [{'GENE': 'MET',\n",
       "   'AFFECTED EXON': '14/21',\n",
       "   'BREAKPOINT 1': 'chr7:116411708',\n",
       "   'BREAKPOINT 2': 'chr7:116414934',\n",
       "   'SPLICE SUPPORTING READS': 3525}])"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a8832e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNVs & Indels (Clinical Significance)\n",
    "vcs_data = SNV[SNV['Clinical_significance'] == 'VCS'][['Gene', 'Consequence', 'AA Change', 'VAF', 'HGVSc', 'HGVSp']]\n",
    "vcs_data.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0ea11f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fusion Gene (Clinical Significance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd4af7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy number variation (Clinical Significance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5305a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Large rearrangements in BRCA1/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8567a961",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Splice Variant (Clinical Significance)\n",
    "splice_vcs = Splice[Splice['Clinical_significance'] == 'VCS']\n",
    "\n",
    "Splice_Variant_Highlight = ', '.join(splice_vcs[\"highlight\"].tolist())\n",
    "Splice_Variant_list = []\n",
    "\n",
    "if not Splice_Variant_Highlight:\n",
    "    Splice_Variant_Highlight = None\n",
    "else:    \n",
    "    for idx, row in splice_vcs.iterrows():\n",
    "        Splice_Variant_list.append({\n",
    "            \"GENE\": row[\"Gene\"],\n",
    "            \"AFFECTED EXON\": row[\"Affected exon\"],\n",
    "            \"BREAKPOINT 1\": row[\"Breakpoint 1\"],\n",
    "            \"BREAKPOINT 2\": row[\"Breakpoint 2\"],\n",
    "            \"SPLICE SUPPORTING READS\": row[\"Splice supporting reads\"]\n",
    "        })\n",
    "        \n",
    "Splice_Variant_Highlight, Splice_Variant_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "354074b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metrics</th>\n",
       "      <th>Value</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Total TMB (mut/Mb)</td>\n",
       "      <td>13.3</td>\n",
       "      <td>Number of Passing Eligible Variants / \\nCoding...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Number of Passing Eligible Variants</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Coding Region Size in Megabases</td>\n",
       "      <td>1.27</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TMB categorical result</td>\n",
       "      <td>intermediate</td>\n",
       "      <td>High &gt; 20, Low &lt; 10, or Intermediate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TMB highlight</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TMB comment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Percent Unstable MSI Sites (%)</td>\n",
       "      <td>3.15</td>\n",
       "      <td>Total MSI Sites Unstable / \\nUsable MSI Sites ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Total MSI Sites Unstable</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Usable MSI Sites</td>\n",
       "      <td>127</td>\n",
       "      <td>Guideline &gt; 40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MSI categorical result</td>\n",
       "      <td>low</td>\n",
       "      <td>High &gt; 20, Low &lt; 10, or Intermediate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MSI highlight</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MSI comment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Metrics         Value  \\\n",
       "0                    Total TMB (mut/Mb)          13.3   \n",
       "1   Number of Passing Eligible Variants            17   \n",
       "2       Coding Region Size in Megabases          1.27   \n",
       "3                TMB categorical result  intermediate   \n",
       "4                         TMB highlight           NaN   \n",
       "5                           TMB comment           NaN   \n",
       "6                                   NaN           NaN   \n",
       "7        Percent Unstable MSI Sites (%)          3.15   \n",
       "8              Total MSI Sites Unstable             4   \n",
       "9                      Usable MSI Sites           127   \n",
       "10               MSI categorical result           low   \n",
       "11                        MSI highlight           NaN   \n",
       "12                          MSI comment           NaN   \n",
       "\n",
       "                                                 Note  \n",
       "0   Number of Passing Eligible Variants / \\nCoding...  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                High > 20, Low < 10, or Intermediate  \n",
       "4                                                 NaN  \n",
       "5                                                 NaN  \n",
       "6                                                 NaN  \n",
       "7   Total MSI Sites Unstable / \\nUsable MSI Sites ...  \n",
       "8                                                 NaN  \n",
       "9                                      Guideline > 40  \n",
       "10               High > 20, Low < 10, or Intermediate  \n",
       "11                                                NaN  \n",
       "12                                                NaN  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "133a1b45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Estimated Tumor Mutation Burden (TMB)': 13.3,\n",
       " 'Microsatellite Instability (MSI) (% Unstable MSI Sites)': 3.15}"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Other Biomarkers\n",
    "{\n",
    "    'Estimated Tumor Mutation Burden (TMB)': IO['Value'][0],\n",
    "    'Microsatellite Instability (MSI) (% Unstable MSI Sites)': IO['Value'][7],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58fb357f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNVs & Indels (Unknown Significance)\n",
    "vus_data = SNV[SNV['Clinical_significance'] == 'VUS'][['Gene', 'Consequence', 'AA Change', 'VAF', 'HGVSc', 'HGVSp']]\n",
    "vus_data.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94766faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fusion Gene (Unknown Significance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf5d331b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy number variation (Unknown Significance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e50a40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Splice Variant (Unknown Significance)\n",
    "splice_vcs = Splice[Splice['Clinical_significance'] == 'VUS']\n",
    "\n",
    "Splice_Variant_Highlight = ', '.join(splice_vcs[\"highlight\"].tolist())\n",
    "Splice_Variant_list = []\n",
    "\n",
    "if not Splice_Variant_Highlight:\n",
    "    Splice_Variant_Highlight = None\n",
    "else:    \n",
    "    for idx, row in splice_vcs.iterrows():\n",
    "        Splice_Variant_list.append({\n",
    "            \"GENE\": row[\"Gene\"],\n",
    "            \"AFFECTED EXON\": row[\"Affected exon\"],\n",
    "            \"BREAKPOINT 1\": row[\"Breakpoint 1\"],\n",
    "            \"BREAKPOINT 2\": row[\"Breakpoint 2\"],\n",
    "            \"SPLICE SUPPORTING READS\": row[\"Splice supporting reads\"]\n",
    "        })\n",
    "        \n",
    "Splice_Variant_Highlight, Splice_Variant_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b6c50b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Failed gene\n",
    "failed_gene = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "178cadf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 코멘트 파트\n",
    "\n",
    "SNV_Comments = SNV[\"Comment\"].dropna().tolist()\n",
    "CNV_Comments = CNV[\"Comment\"].dropna().tolist()\n",
    "CNVarm_Comments = CNVarm[\"Comment\"].dropna().tolist()\n",
    "Fusion_Comments = Fusion[\"Comment\"].dropna().tolist() \n",
    "Splice_Comments = Splice[\"comment\"].dropna().tolist()\n",
    "Comments = '\\n\\n'.join(SNV_Comments+CNV_Comments+CNVarm_Comments+Fusion_Comments+Splice_Comments)\n",
    "Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8eb2ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 검사정보\n",
    "검사기기 = NGS_QC[4][1]\n",
    "if panel == 'GE':\n",
    "    aa = {\n",
    "        \"검사시약\":\"AllPrep DNA/RNA FFPE Kit (50) [Qiagen], TruSight™ Oncology 500 kit [Illumina]\",\n",
    "        \"검사방법\":\"NGS targeted DNA/RNA sequencing (Library : Hybrid capture)\",\n",
    "        \"검사기기\":검사기기,\n",
    "        \"Reference genome\": \"Homo_sapiens/ UCSC/ hg19\"\n",
    "    }\n",
    "else: # SA\n",
    "    aa = {\n",
    "        \"검사시약\":\"AllPrep DNA/RNA FFPE Kit (50) [Qiagen], TruSight™ Oncology 500 kit [Illumina], TruSight™ RNA Fusion Panel [Illumina]\",\n",
    "        \"검사방법\":\"NGS targeted DNA/RNA sequencing (Library : Hybrid capture)\",\n",
    "        \"검사기기\":검사기기,\n",
    "        \"Reference genome\": \"Homo_sapiens/ UCSC/ hg19\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "b7aecc57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter history\n",
    "filter_history = {'Include': 'Exonic, Illumina Q.C Filter PASS, Fold change <0.5 or >1.5', 'Exclude': 'Synonymous, VAF <3%, total depth <100, refer depth=0'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "0ea25f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DNA, RNA (Qubit 농도)\n",
    "QubitValue = {\n",
    "    'DNA': str(clinical_dict[\"DNA conc.(ng/ul)\"]),\n",
    "    'RNA': str(clinical_dict[\"RNA conc.(ng/ul)\"])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "9f5a907e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Metric (UOM)', 'LSL Guideline', 'Value'],\n",
       " ['PCT_PF_READS (%)', 80, 85.1],\n",
       " ['PCT_Q30_R1 (%)', 80, 94.8],\n",
       " ['PCT_Q30_R2 (%)', 80, 94.3]]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Q.C\n",
    "QC_Value = [\n",
    "    [\"Metric (UOM)\", 'LSL Guideline', 'Value'],\n",
    "    [NGS_QC[1][7], 80, NGS_QC[3][7]],\n",
    "    [NGS_QC[1][8], 80, NGS_QC[3][8]],\n",
    "    [NGS_QC[1][9], 80, NGS_QC[3][9]],\n",
    "]\n",
    "QC_Value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "e1f587a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analysis Program\n",
    "analsys_program = \"DRAGEN TSO500 ( Workflow Version : 2.5.2 )\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36baf15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specification\n",
    "if panel == 'GE':\n",
    "    specification = {\n",
    "        'Parameter': 'TruSight Oncology 500',\n",
    "        'System': 'NextSeq 550Dx or NovaSeq 6000Dx (research mode)',\n",
    "        'Panel Size': '1.94 Mb DNA, 358 kb RNA',\n",
    "        'DNA input requirement': '40 ng total',\n",
    "        'RNA input requirement': '40 ng total',\n",
    "        'FFPE input requirement': 'Minimum recommendation of 0.65㎣ from FFPE tissue samples (5 slides cut at 10 microns thick)',\n",
    "        'Total assay time': '3-4 days from nucleic acid to variant report',\n",
    "        'Sequene run time': '24 hours',\n",
    "        'Sequene run': '2 X 101 cycles',\n",
    "        'Kit size': '24 or 48 samples',\n",
    "        'Sample throughput': '8 or 16 or 32 samples per run',\n",
    "        'Limit of detection': '5% VAF for small variants 5 copies per ng RNA input for fusions 1.5x fold-change for CNVs',\n",
    "        'Analytical sensitivity': '>96% (for all variant types at 5% VAF)',\n",
    "        'Analytical specificity': '>99.99%'\n",
    "    }\n",
    "else: # SA\n",
    "    specification = {\n",
    "        'Parameter': 'TruSight Oncology 500 / TruSight RNA Fusion Panel',\n",
    "        'System': 'NextSeq 550Dx or NovaSeq 6000Dx (research mode)',\n",
    "        'Panel Size': '1.94 Mb DNA, 358 kb RNA',\n",
    "        'DNA input requirement': '40 ng total',\n",
    "        'RNA input requirement': '40 ng total',\n",
    "        'FFPE input requirement': 'Minimum recommendation of 0.65㎣ from FFPE tissue samples (5 slides cut at 10 microns thick)',\n",
    "        'Total assay time': '3-4 days from nucleic acid to variant report',\n",
    "        'Sequene run time': '24 or 19 hours',\n",
    "        'Sequene run': '2 X 101 cycles',\n",
    "        'Kit size': '24 or 48 samples',\n",
    "        'Sample throughput': '8 or 16 or 32 samples per run',\n",
    "        'Limit of detection': '5% VAF for small variants 5 copies per ng RNA input for fusions 1.5x fold-change for CNVs',\n",
    "        'Analytical sensitivity': '>96% (for all variant types at 5% VAF)',\n",
    "        'Analytical specificity': '>99.99%'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "4ea1f604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tested, Signed, Analyzed by, 분자 접수 번호\n",
    "tested_by = ', '.join([clinical_dict[i] for i in list(clinical_dict.keys()) if i.startswith('Tester')])\n",
    "signed_by = ', '.join([clinical_dict[i] for i in list(clinical_dict.keys()) if i.startswith('Signed')])\n",
    "analyzed_by = '이청'\n",
    "분자접수번호 = clinical_dict[\"분자접수번호\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
